### PR TITLE
[MSHARED-843] Implement Maven dependency collector to wrap Aether fun…

### PR DESCRIPTION
**Pull request by pmoerenhout**
Implement Maven dependency collector to wrap Aether functionality. If this is merged, MDEP-648, can be implemented.

### CODE REVIEW

File: src/main/java/org/apache/maven/shared/transfer/collection/internal/Maven30ArtifactRepositoryAdapter.java [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+    {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public String toString()
+    {
+        StringBuilder sb = new StringBuilder();
+
+        sb.append( "       id: " ).append( getId() ).append( "\n" );
```
Comment (michael-o) [inline src/main/java/org/apache/maven/shared/transfer/collection/internal/Maven30ArtifactRepositoryAdapter.java]: Use `System.lineSeparator()` throughout.

File: src/main/java/org/apache/maven/shared/transfer/collection/internal/Maven31ArtifactRepositoryAdapter.java [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+    {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public String toString()
+    {
+        StringBuilder sb = new StringBuilder();
+
+        sb.append( "       id: " ).append( getId() ).append( "\n" );
```
Comment (michael-o) [inline src/main/java/org/apache/maven/shared/transfer/collection/internal/Maven31ArtifactRepositoryAdapter.java]: Same here.

File: src/main/java/org/apache/maven/shared/transfer/dependencies/collect/internal/Maven30DependencyNodeAdapter.java [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+        return dependencyNode.getDependency().getScope();
+    }
+
+    @Override
+    public boolean accept( DependencyVisitor visitor )
+    {
+        if ( visitor.visitEnter( this ) )
+        {
+            List<org.sonatype.aether.graph.DependencyNode> aetherNodes = dependencyNode.getChildren();
+            for ( int i = 0, n = aetherNodes.size(); i < n; i++ )
```
Comment (michael-o) [inline src/main/java/org/apache/maven/shared/transfer/dependencies/collect/internal/Maven30DependencyNodeAdapter.java]: Why not for each?

File: src/main/java/org/apache/maven/shared/transfer/dependencies/collect/internal/Maven31DependencyNodeAdapter.java [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+        return dependencyNode.getDependency().getOptional();
+    }
+
+    @Override
+    public boolean accept( DependencyVisitor visitor )
+    {
+        if ( visitor.visitEnter( this ) )
+        {
+            List<org.eclipse.aether.graph.DependencyNode> aetherNodes = dependencyNode.getChildren();
+            for ( int i = 0, n = aetherNodes.size(); i < n; i++ )
```
Comment (michael-o) [inline src/main/java/org/apache/maven/shared/transfer/dependencies/collect/internal/Maven31DependencyNodeAdapter.java]: Same here.

**Review by michael-o — CHANGES_REQUESTED** [review]
Don't throw IAE on null input, it should be a NPE.

**Comment by pmoerenhout** [comment]
> Don't throw IAE on null input, it should be a NPE.

]

> Don't throw IAE on null input, it should be a NPE.

I added the use of Objects.requireNonNull for the new code.

File: src/main/java/org/apache/maven/shared/transfer/collection/internal/Maven30ArtifactRepositoryAdapter.java (hunk 1 above)
Comment (pmoerenhout) [reply src/main/java/org/apache/maven/shared/transfer/collection/internal/Maven30ArtifactRepositoryAdapter.java]: Used the private static final String LS = System.lineSeparator();

File: src/main/java/org/apache/maven/shared/transfer/dependencies/collect/internal/Maven30DependencyNodeAdapter.java (hunk 3 above)
Comment (pmoerenhout) [reply src/main/java/org/apache/maven/shared/transfer/dependencies/collect/internal/Maven30DependencyNodeAdapter.java]: Indeed, replaced with for().

Comment (rfscholte) [reply src/main/java/org/apache/maven/shared/transfer/dependencies/collect/internal/Maven30DependencyNodeAdapter.java]: I'm pretty sure @michael-o was expecting this:

    for ( org.sonatype.aether.graph.DependencyNode child : dependencyNode.getChildren() )
    {
         DependencyNode node = new Maven30DependencyNodeAdapter( child );
         if ( !node.accept( visitor ) )
         {
             break;
         }
    } 

Easier to read and probably more efficient.

Comment (pmoerenhout) [reply src/main/java/org/apache/maven/shared/transfer/dependencies/collect/internal/Maven30DependencyNodeAdapter.java]: Implemented the expected!

**Pull request closed without merging**